### ISSUE TITLE
How to group nested async actions in a single transaction?

### ISSUE THREAD

**Issue by nikolovp** [mobxjs/mobx#1597]
Hi, I have been using mobx for some time but my actions have never done anything async so I didn't have problems with nesting actions in a single transaction until now.

Now I have async actions that perform multiple async operation where each async operation could be another async action that modifies some observable data. What I need is to group all data updates in a single transaction no matter in which action the changes were made.

I have read the doc for async actions but I still cannot figure out how to deal with multiple nested async actions.

Simplified example based on MobX 5: https://jsfiddle.net/ppetrov/b01kwuhx/

I haven't tried generators because I moved from generators to async/await an year ago and it fills like  a step backwards.

**Comment by mweststrate** [comment]
There is deliberately no concept of *async transactions* in mobx (you can search the issue tracker on the why not, it has been asked quite a few times before). So basically there are two approaches

1. only update your state at the end of your async process, instead of in the middle in it. (this can be achieved easily by restructuring your actions or for example keeping a list of thunks that need to be applied once the whole process is finished)
2. update your state in the middle of the process, just make sure that every code is wrapped in an action. `flow` does this automatically for you, but if you don't like generators you can also combine `async / await` with `@action` or `runInAction`

This page explains several of the different approaches: https://mobx.js.org/best/actions.html

**Comment by nikolovp** [comment]
First of all, thanks for the quick response.

I am more after the second option because I my case each async action can be also executed separately.

I prefer `async / await` vs generators because you cannot execute generator without an yield and if you miss the yeld you do not get any kind of error. It is annoyingly difficult to track issues caused by not executed generators.

In the JSFiddle example I provided all async operations are already defined as actions but the reactions still get fired on each data update instead of when the outermost action is done. I also tried with `runInAction` without much success.

I already looked for similar issues and the described approaches in https://mobx.js.org/best/actions.html but I didn't see anywhere a working example with more than one nested async action.

Any ideas how to make the JSFiddle example work? I will also try with generators.

**Comment by kuitos** [comment]
@nikolovp try https://github.com/mobxjs/babel-plugin-mobx-deep-action and https://github.com/Strate/babel-plugin-mobx-async-action for `async/await`

**Comment by nikolovp** [comment]
@kuitos Thanks for the links. I found those two babel plugins yesterday but since all my async operations are already actions I do not see difference it will make. I will give them a try anyway.

**Comment by nikolovp** [comment]
The `async / await` example is updated: https://jsfiddle.net/ppetrov/b01kwuhx/
The same example but with generators and `mobx.flow`: https://jsfiddle.net/ppetrov/qd6t9gxz/

With generators I get exactly the same results as with `async / await`. Am I missing something fundamental in those examples?

**Comment by mweststrate** [comment]
I think I might misunderstand what you are trying to achieve. What is your goal of having an 'async transaction'. That changes are not made until the action finishes? Such a concept does not exist in MobX, as it doesn't (can't nor won't) fork your entire state to achieve such a thing (it would be extremely complicated to reason about)

**Comment by kuitos** [comment]
@nikolovp check https://github.com/mobxjs/mobx/issues/1206 and https://github.com/mobxjs/mobx/issues/820

**Comment by mweststrate** [comment]
Yes, that is not possible. MobX always produces a consistent state at the end of 'tick'. Not running reactoins would means your application is inconsistent, for example that your UI is rendering data which doesn't  match what is in your state. If  you don't want to trigger certain reactions, you have to postpone your state updates. Without that, there is no way to reason about your application. take for example the following super simple example:

```
const x = observable({ a: 1, b: 2 })

autorun(() => renderInUI(x.a + x.b)

setTimeout(() => {
   x.b += 1
}, 100)

async action() {
  delay 50
  x.a = 4
  delay 100
  x.a = 5
}
```
No semantics will become very vague. Should autorun calculate with the original `a` or the new one? Or should it not run at all after changing `b` because `a` is still in transaction? And once you are debugging, why is my state updated but is my UI stale? There might be a pending promise *somewhere* in the app, causing your entire app to freeze, but how to keep track of such pending promises...

So if the intermediate state of x.a is not supposed to become visible, just don't update that state there, or create a list of things todo: 

```
async action() {
  const tasksLeft = []
  delay 50
  tasksLeft.push(() => x.a = 4)
  delay 100
  tasksLef.push(() => x.a = 5)
  // done!
  tasks.forEach(f => f())
}
```

The reasoning is simple: in MobX there is only single source of truth: the state. And everything is always consistent with that state. So after each state change all reactions are fired. Yes this is a bit delayed, but that is fine as reactoins are only supposed to to trigger side effects, which effects are *invisible* from both code and user perspective. However making transactions over time would make these inconsistencies (a not yet propagated change) detectable by both code and users.

**Comment by nikolovp** [comment]
Oh, after you explained it that way, what I want to achieve is conceptually wrong indeed. I guess I missunderstood the nesting of actions and the async actions.
@mweststrate @kuitos Thanks for the help and clarification!

**Comment by mweststrate** [comment]
No problem!

Op di 19 jun. 2018 om 13:02 schreef Petar Petrov <notifications@github.com>:
